### ISSUE TITLE
Migrate LogCollect CASTOR endpoint to CTA

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#10110]
**Impact of the new feature**
WMAgent

**Is your feature request related to a problem? Please describe.**
CASTOR has been recently put in read-only mode and it's getting decommissioned in favor of the CERN CTA tape technology.
That means WMAgent components that depend on CERN CASTOR have also to be migrated/adapted to this new technology.

**Describe the solution you'd like**
Discussion with experts is still in the initial phase, however some of the information provided by Katy (and her contacts at CERN are):
```
The production instance name will be eosctacms.cern.ch
The Rucio prefix should be /eos/ctacms/archive but I will confirm when the final import is finished.
So the PFN would be : root://eosctacms.cern.ch//eos/ctacms/archive
```

It's still not clear whether `xrdcp` will work and what authentication/authorization mechanism will be required. So this needs further discussion.

**Describe alternatives you've considered**
Another alternative suggested by Eric V. would be to manage those logs with Rucio, or perhaps use the `rucio upload` mechanism to write these files to the CTA storage. Another topic that requires further discussion.

**Additional context**
Just to add some context, the current behaviour of such LogCollect jobs is the following:
1) if possible, use `edmCopyUtil` from the CMSSW release (thus CMSSW needs to be bootstrapped) to stage files in
2) copy files from /store/unmerged/logs into the worker node scratch area
3) add all the successful logs in the scratch area to a big tarball (delete every one of them as soon as they get added to the tarball)
4) write the tarball to CASTOR (/store/logs/prod), abort the job if any failure happens
5) try to write the tarball CERN EOS in a best effort mode (failures are discarded)
6) last but not least, delete every single file that we managed to copy from the /store/unmerged/logs to the scratch area. These are deleted from /store/unmerged/logs

FIXME TODO:
While looking at the code, there seems to be a bug on the file deletion from the scratch area (not every log is deleted, but only the last one in the loop):
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/Steps/Executors/LogCollect.py#L204

and a second problem - or improvement - would be to delete all the logArchive files passed to this job, not only those that we managed to copy to the scratch area:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/Steps/Executors/LogCollect.py#L262

**Comment by amaltaro** [comment]
Just an update on this issue. I have performed some interactive tests with `rucio upload`, also discussed a few things with Katy and Eric, here is a summary and the current status.

Eric has created a new scope, called `logs`, where we can write these logs to.
A new Rucio account has also been created for these activities, the account name is `wma_logs`.
Last but not least, given that grid jobs run with my personnel proxy, Eric mapped my personnel DN to the `wma_logs` rucio account.

Here is the outcome of the interactive tests I was running on lxplus.
Setting up the rucio client environment:
```
source /cvmfs/cms.cern.ch/cmsset_default.sh
voms-proxy-init -voms cms -rfc -valid 192:00
source /cvmfs/cms.cern.ch/rucio/setup.sh
export RUCIO_ACCOUNT=wma_logs
```

then accessing the directory with the log tarball, and writing it to CERN EOS:
```
amaltaro@lxplus783:~/workspace $ rucio -v upload --scope logs --rse T2_CH_CERN --lifetime 86400  --name /store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar LogCollect-cmswn2292-1-logs.tar
...
2021-01-13 15:08:12,391	INFO	Preparing upload for file LogCollect-cmswn2292-1-logs.tar
...
2021-01-13 15:08:12,572	INFO	Successfully added replica in Rucio catalogue at T2_CH_CERN
2021-01-13 15:08:12,689	INFO	Successfully added replication rule at T2_CH_CERN

...
2021-01-13 15:08:14,253	INFO	Trying upload with root to T2_CH_CERN
2021-01-13 15:08:14,253	DEBUG	uploadclient.py	Processing upload with the domain: wan
2021-01-13 15:08:14,439	DEBUG	uploadclient.py	The PFN created from the LFN: root://eoscms.cern.ch:1094//eos/cms/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar
2021-01-13 15:08:14,763	DEBUG	utils.py	put: Attempt 1
2021-01-13 15:08:15,023	INFO	Successful upload of temporary file. root://eoscms.cern.ch:1094//eos/cms/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar.rucio.upload
2021-01-13 15:08:15,023	DEBUG	uploadclient.py	skip_upload_stat=False
2021-01-13 15:08:15,024	DEBUG	uploadclient.py	stat: pfn=root://eoscms.cern.ch:1094//eos/cms/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar.rucio.upload
2021-01-13 15:08:15,026	DEBUG	uploadclient.py	Filesize: Expected=225280 Found=225280
2021-01-13 15:08:15,027	DEBUG	uploadclient.py	Checksum: Expected=675eedbd Found=675eedbd
2021-01-13 15:08:15,027	DEBUG	uploadclient.py	Renaming file root://eoscms.cern.ch:1094//eos/cms/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar.rucio.upload to root://eoscms.cern.ch:1094//eos/cms/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar
2021-01-13 15:08:15,030	DEBUG	uploadclient.py	Upload done.
2021-01-13 15:08:15,031	INFO	Successfully uploaded file LogCollect-cmswn2292-1-logs.tar
```

few observations are:
* the file gets properly renamed to whatever we pass in the end of the `--name` parameter
* rucio upload first adds that file replica to the Rucio catalogue, then it makes the transfer
* `--lifetime 86400` is provided in this test. Help pages say it's the lifetime of the rule(?) Did we also get a rule created here? Is the file eligible for deletion after that lifetime?

What's left to be investigated is how we can create a container (or block) and attach these replicas to them. The agent does not keep track of the output of LogCollect jobs, so making such operation on the agent side would be quite intrusive with many changes.

@ericvaandering @KatyEllis FYI

**Comment by ericvaandering** [comment]
Yeah, there is a rule: -bash-4.2$ rucio list-rules logs:/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar 
ID                                ACCOUNT    SCOPE:NAME                                                                                                                           STATE[OK/REPL/STUCK]    RSE_EXPRESSION      COPIES  EXPIRES (UTC)        CREATED (UTC)
--------------------------------  ---------  -----------------------------------------------------------------------------------------------------------------------------------  ----------------------  ----------------  --------  -------------------  -------------------
4a7f0397a3f34ca0a6b13214c6bb4bdc  wma_logs   logs:/store/logs/prod/2021/01/WMAgent/amaltaro_TaskChain_ProdMinBias_Jan2021_Val_210112_212625_3216/LogCollect-cmswn2292-2-logs.tar  OK[1/0/0]               T2_CH_CERN               1  2021-01-14 14:08:12  2021-01-13 14:08:12


Shall we coordinate by Slack with setting up a subscription which copies these to CTA?

**Comment by amaltaro** [comment]
Another update to this thread, after what we discussed in yesterday's Rucio transition meeting.
In addition to running the `rucio upload` command to transfer (and register) the log tarball to the CERN EOS storage, we will also try to create a container DID and attach the tarball file to it. These 3 operations will all be done from the worker node.

Container DID will likely be named after the workflow name, thus we should expect and properly handle the `DataIdentifierAlreadyExists` exception.

**Comment by amaltaro** [comment]
Defined as low prio, from the point of view of the P&R team (Hasan)

**Comment by amaltaro** [comment]
And from the question raised during the O&C weekly meeting, here is a breakdown of how much space these logs are meant to take in CERN CTA:
```
amaltaro@vocms0275:~ $ du -sh /eos/cms/store/logs/prod/2022/01/WMAgent/
7.3T	/eos/cms/store/logs/prod/2022/01/WMAgent/
amaltaro@vocms0275:~ $ du -sh /eos/cms/store/logs/prod/2022/02/WMAgent/
6.1T	/eos/cms/store/logs/prod/2022/02/WMAgent/
amaltaro@vocms0275:~ $ du -sh /eos/cms/store/logs/prod/2022/03/WMAgent/
4.9T	/eos/cms/store/logs/prod/2022/03/WMAgent/
```

considering the past 3 months. This EOS area has exactly the same tarballs that we would be trying to write to CTA.

**Comment by amaltaro** [comment]
Having not heard any requests to get this feature ready for data taking, and from a feedback from Danilo that the current data volume is okay to be kept in EOS, I am downgrading it (removing the Data taking label) and we can work on it in the quarters to come.
